### PR TITLE
Configurable UI branding, keyboard list and hide-able columns through a new config.js file

### PR DESCRIPTION
**Pull request by anuragaw**
## Description

We want to support hiding table columns, specifically in metrics table, through config file so that users can make the relevant bits hidden as per their organization. Current work will support the metrics table but can be extended to any table with minimal work in future.

Config file will take the key of the metrics column from metrics.js file for the sake of minimal changes and simplicity of development.

**Problem**: The keyboard list in the UI is not consistent across views such as in the instance wizard and in the register template form. There is also no way to custom about url/text and doc title and help URL in the UI.
**Root Cause**: The list is hardcoded in the UI allowing no centralised configuration.
**Solution**: Introduce a new `config.js` file installed at the `/usr/share/cloudstackmanagement/webapp/config.js` location. The config.js allows configurable keyboard list, about url/text, doc title,  and help URL.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?
Testing and further documentation in progress and will be updated with proper notes.

Image with all columns showing up -
unhiddencolumn

Change in config.js file -
config

View after change in config file -
hiddencolumn

Another Example with Hosts Metrics - 

hosts_allvisible

hosts_config

hosts_hidden

### CODE REVIEW

**Comment by rhtyd** [comment]
At some point, let's merge this with PR https://github.com/apache/cloudstack/pull/3217 otherwise this will lead to merge conflicts for related features extending/using config.js

**Comment by anuragaw** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
Put some after-loaded method/js file that does something looping over sections minimizing changes to the listview/widget/library:
```
            $.each(cloudStack.sections, function(idx, section) {                   
                if (section.listView) {                                            
                    section.listView.preFilter = function(args) {                  
                        const original = section.listView.preFilter(args);         
                        return ['name'] + original;                                                                                                                     
                    };                                                             
                    console.log(Object.keys(section.listView.fields));             
                }                                                                  
            }); 
```

**Comment by rhtyd** [comment]
Closed https://github.com/apache/cloudstack/pull/3217 in favour of this one

File: ui/scripts/ui/widgets/listView.js [hunk 1]
```diff
@@ -860,6 +860,8 @@
         if (groupableColumns) {
             $tr.addClass('groupable-header-columns').addClass('groupable-header');
             $.each(fields, function(key) {
```
Comment (anuragaw) [inline ui/scripts/ui/widgets/listView.js]: This was broken in existing code and needed fixing. If an upper field is hidden we must not draw it's containing columns.

**Comment by anuragaw** [comment]
@rhtyd I have updated the PR with the changes we discussed.

**Comment by borisstoyanov** [comment]
@blueorangutan package

File: ui/config.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Define custom options configurable by admins for UI
+cloudStackOptions = {
+    aboutText: "label.app.name",
+    aboutTitle: "label.about.app",
+    docTitle: "label.app.name",
```
[the remark this answers is not in the document]
Comment (anuragaw) [reply ui/config.js]: Done

File: ui/config.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Define custom options configurable by admins for UI
+cloudStackOptions = {
+    aboutText: "label.app.name",
```
[the remark this answers is not in the document]
Comment (anuragaw) [reply ui/config.js]: Done

**Comment by anuragaw** [comment]
@PaulAngus - addressed all the comments in the latest commit. There are 5 keyboards specified in the config.js. I get first 4 of them. I wanted to confirm if the keyboard options you get were only missing "Simplified Chinese Keyboard" or any more?

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
Purely UI changes, merging based on tests and reviews.

**Comment by anuragaw** [comment]
Created PR https://github.com/apache/cloudstack/pull/3429 to clean up config file with redundant columns.

**Pull request merged**